src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+        _ => (Maybe, Maybe),
+    }
+}
+
+fn variadic_monotonic(func: &VariadicFunc) -> Monotonic {
+    use Monotonic::*;
+    use VariadicFunc::*;
+    match func {
+        And | Or => Yes,
+        _ => Maybe,
+    }
+}
+
+/// An inclusive range of non-null datum values.
+#[derive(Clone, Eq, PartialEq, Debug)]
+enum Values<'a> {
+    /// This range contains no values.
+    Empty,
+    /// An inclusive range. Invariant: the first element is always <= the second.
+    Within(Datum<'a>, Datum<'a>),
+    /// Constraints on structured fields, useful for recursive structures like maps.
+    /// Fields that are not present in the map default to Values::All.
+    Nested(BTreeMap<Datum<'a>, ResultSpec<'a>>),
+    /// This range might contain any value. Since we're overapproximating, this is often used
+    /// as a "safe fallback" when we can't determine the right boundaries for a range.
+    All,
+}
+
+impl<'a> Values<'a> {
+    fn just(a: Datum<'a>) -> Values<'a> {
+        match a {
+            Datum::Map(datum_map) => Values::Nested(
+                datum_map
+                    .iter()
+                    .map(|(key, val)| (key.into(), ResultSpec::value(val)))
+                    .collect(),
+            ),
+            other => Self::Within(other, other),
+        }
+    }
+
+    fn union(self, other: Values<'a>) -> Values<'a> {
+        match (self, other) {
+            (Values::Empty, r) => r,
+            (r, Values::Empty) => r,
+            (Values::Within(a0, a1), Values::Within(b0, b1)) => {
+                Values::Within(a0.min(b0), a1.max(b1))
+            }
+            (Values::Nested(mut a), Values::Nested(mut b)) => {
+                a.retain(|datum, values| {
+                    if let Some(other_values) = b.remove(datum) {
+                        *values = values.clone().union(other_values);
+                    }
+                    *values != ResultSpec::anything()
+                });
+                if a.is_empty() {
+                    Values::All
+                } else {
+                    Values::Nested(a)
+                }
+            }
+            _ => Values::All,
+        }
+    }
+
+    fn intersect(self, other: Values<'a>) -> Values<'a> {
+        match (self, other) {
+            (Values::Empty, _) => Values::Empty,
+            (_, Values::Empty) => Values::Empty,
+            (Values::Within(a0, a1), Values::Within(b0, b1)) => {
+                let min = a0.max(b0);
+                let max = a1.min(b1);
+                if min <= max {
+                    Values::Within(min, max)
+                } else {
+                    Values::Empty
+                }
+            }
+            (Values::Nested(mut a), Values::Nested(b)) => {
+                for (datum, other_spec) in b {
+                    let spec = a.entry(datum).or_insert_with(ResultSpec::anything);
+                    *spec = spec.clone().intersect(other_spec);
+                }
+                Values::Nested(a)
```

Comment:
If we're over-approximating and one of the fields is empty, would that mean the whole map is empty?

Reply:
On reflection: yes!

(I'm not sure how much effort to put into normalizing stuff like this... any takes?)

Reply:
Identifying impossible paths can be a huge win, so I'd say it's worth doing at some point. Whether it should be done here or in the optimizer itself I'm less clear on---from the optimizer's perspective, it's a missed opportunity if we're generating code with impossible paths.

Relatedly, something I'm not clear on: when does this analysis happen? Before running the optimizer, at the end, or at some point in the middle?

Reply:
FYI, this is the issue that we have for making the impossible condition detection in the optimizer better: https://github.com/MaterializeInc/materialize/issues/18091